base/special/trig.jl

```diff
@@ -860,8 +860,8 @@ function cospi(x::T) where T<:Union{Integer,Rational}
     end
 end
 
-sinpi(x::Integer) = x >= 0 ? zero(float(x)) : -zero(float(x))
-cospi(x::Integer) = isodd(x) ? -one(float(x)) : one(float(x))
+sinpi(x::Integer) = 0
+cospi(x::Integer) = isodd(x) ? -1 : 1
```

Comment:
Same here: `one(x)` or `-one(x)`?

Reply:
That wouldn't work for `x::Unsigned`, so we'd most definitely want `±one(Signed(x))`. The reason I just chose `Int` literals is that the input range won't have any effect on the output range, which will always be `±1`, so it won't make sense to use a whole `BigInt` just because the input is `BigInt`, for example. I thought about always returning `Int8` here, but that just somehow felt wrong, since it's really uncommon for functions to return `Int8` if one of the inputs isn't `Int8`.

Reply:
An argument against `Int8` is also that it can overflow really quickly if you do any calculations with it, so we probably don't want to return it if the user isn't expecting an `Int8`

Reply:
I'm with you for `Unsigned`, though for signed integers I think people know what they are doing if they plug in smaller integer types than `Int64`. Assume I want to keep memory load on the registers as low as my use case allows, so I plug in `Int16` instead of `Int64` and I don't want any computation to promote types to `Int64`. So I would need to do `Int16(f(i16))`. I don't know whether the compiler would optimize the conversion roundtrips but I doubt it. Hence it's not only clumsy but also inefficient.

`f` could be any function, not necessarily `sinpi` or `cospi`. Assume it's from a package of which I don't know the internals. Then I don't even know at which point computations "exploded" into the wrong data type. Therefore I'd say this should be `Signed(one(x))` or `-Signed(one(x))` - `Signed` at the outside, so that it only changes the type but not the bit representation.

Reply:
Ok, that makes sense.
> Therefore I'd say this should be Signed(one(x)) or -Signed(one(x)) - Signed at the outside, so that it only changes the type but not the bit representation.
I don't quite understand what you mean here. I think we can just call `Signed` on the type of `x` though, which should eliminate all unnecessary allocations, e.g. with `BigInt`.

Reply:
`one(Signed(x))` performs to operations: first convert `x` to its associated signed type (which might fail: `Signed(typemax(UInt8))`), then construct the 1 of that type (discarding the reinterpreted `x`).

`Signed(one(x))` essentially performs only one operation: first construct the 1 of `x`'s type (which never fails), then reinterpret it as of a different type (note `bitstring(UInt8(1)) == bitstring(Int8(1))`).

Reply:
I just saw your new implementation. If you go by the types directly, there is no difference. How do I mark this conversation as resolved? (The other one about `sinpi` and `cospi` having matching types remains.)